### ISSUE TITLE
Integration process - review, find pros and cons, think about possible improvements

### ISSUE THREAD

**Issue by segoddnja** [ClearGDPR/ClearGDPR#367]
**Is your feature request related to a problem? Please describe.**
Current integration process is quite cumbersome for Processors. And it causes most challenging problem at this moment - adding node to the cluster with smart contract deployed in private mode.

**Describe the solution you'd like**
Here I compare proposed approach with current one.
![cleargdpr](https://user-images.githubusercontent.com/1170038/43573082-382d05e2-9641-11e8-9470-81c5faebe02f.png)

**Comment by segoddnja** [comment]
I can not completely understand current implementation. Processor has it's own API version running on it's side. This API listens for blockchain events and processes them somehow. For instance when it recieves `Consent given` event, it does HTTP request to controllers API to receive subjects data and create subjects copy in local database. But it can not access this data directly from database because it is encrypted. To access the subjects it needs to make a request to its own API. At the same time Processor can directly interact with controllers API and request any subject data when it is necessary. 
It makes not too much sense to me.

**Comment by Nelrohd** [comment]
I might be very wrong but I think the goal of having a quorum node on each controller/processor is to make the audit log immutable by keeping a copy of it and then being able to compare it with others in order to detect fraud or inconsistency.

**Comment by Nelrohd** [comment]
Client front-end is Element SDK?

**Comment by Nelrohd** [comment]
Also I think the controller has only one quorum node right? @sindelio

**Comment by sindelio** [comment]
Some general comments:

1. I appreciate the effort in trying to make architecture better.
2. In the beginning of the development we agreed to prioritize a simple approach to the processor software, and that was achieved by creating a secondary running mode for it. This simple approach allowed us to continue moving forward at the time, in which we had many unknowns. But, the end goal was to make a separate  and optimized software for the processors. I agree that right now the integration is a little difficult, and we should work on this.
3. We have to bear in mind that the clients want this project to be built on top of Quorum and it's privacy features. So our ideal scenario would be the processors running a Quorum node. 
4. My end-game vision for the project is that we would have a big Quorum network, with more than 1 controller in it, and hence more than 1 private smart contract deployed since they have a 1:1 relation. In this big network, a processor would be able to participate in many private smart contracts, which means a processor could interact with more than 1 controller.

**Comment by segoddnja** [comment]
> I might be very wrong but I think the goal of having a quorum node on each controller/processor is to make the audit log immutable by keeping a copy of it on each node and then being able to compare it with others in order to detect fraud or inconsistency.

As I understand this, all nodes in network are in consistent mode. So, processors node doesn't need to see/use smart contract - it will be kept in consistent state by the quorum mechanism. All interactions with smart contract can(and I think must) happen trough public API hosted by Controller

> Client front-end is Element SDK?

Client front-end is controller's web portal, where data subjects can register, provide personal data, give consent and so on. Elelment SDK is just a library to make integration easier.

**Comment by segoddnja** [comment]
@sindelio 
1-2 Is clear enough.
3. In proposed architecture owning quorum node by Processor is optional, but, it depends on Controller. So, Controller can allow processors to interact with API only after they created this node.
4. I think it is possible only in one case - we provide ClearGDPR solution as SAAS.

**Comment by sindelio** [comment]
Some comments about the architecture:

1. The controller has only one Quorum node in it.
2. We will move forward with a public smart contract, considering a network with only 1 controller for now. We have to make sure this network with a single controller works well before moving to a more complex network with more than 1 controller and private smart contracts.
3. I believe we have 3 cases to address for the processors: a) the processor does not run a Quorum node and does not have a local server and database. In this case it's good that we provide him a server and a database, although we have much work to do in regard to automation and optimization. b) The processor does not run a Quorum node but has it's own local server and database. This will be a common case, specially in the beginning during the integration process itself and popularization of the project. For this case we should make an optimized version of the software for the processor, which will consist basically of an API and communication via webhooks/someting similar with the controller. c) The processor has a Quorum node and local database and server running. This is our ideal scenario. We could also consider the processor having a Quorum node but not having a local database and server, but this is very unlikely.
4. The processor is not free to get any subject's data. He can only fetch data from subjects that he has received consent from. When a processor receives a consent, he automatically fetch the data from the controller, after receiving a consent event in the blockchain network. The processor can also request the data for a subject to the controller, but he will only receive it if he has consent from that subject. To process a subject's data in its own system, a processor needs to call the local API to get back that data in a machine readable format (JSON).

**Comment by segoddnja** [comment]
4. I still can not understand why does processor needs to have a local copy of database with subjects data? I understand concept of quorum network and why every Processor needs to have access to smart contract.So in this case they can see transactions and be sure that everything is transparent. But, why does they need to run own API? @sindelio Can you clarify this?

**Comment by sindelio** [comment]
The database is not required for the processor. Although, having a database makes the *integration* process easier for the processors, because they won't need to set their own tables to store the subjects data. If a processor uses the system as it is, then it can make API calls to fetch data from the database inside the processor instance. If we remove that database, then the processor will need to create his own tables. The tradeoff is a more *complex* application, with a database in it.
The most likely case is that a processor will want the database we made in the processor software. That's because it can then just make API calls to fetch the data, without worrying about creating and managing tables. 
Another case we have to address is when the processor does not run a clearGDPR instance, which means he has no database or quorum node. We were thinking about using webhooks for this case, but we didn't start since it was not a priority.
Let's keep discussing about this, the project will keep evolving

**Comment by segoddnja** [comment]
Well, processor can not run instance without database. To run instance processor must prepare all necessary infrastructure - database and quorum node.

> The most likely case is that a processor will want the database we made in the processor software. That's because it can then just make API calls to fetch the data, without worrying about creating and managing tables.

It makes not much sense to me. Public API provided by Controller for all Processors looks much more convenient. Processor need to call API to get the data in any case. But, it is much easier to just get access token from Controller and start calling public API (and cheaper as you don't need to run any software). Processor may not need to store data on its own, it can fetch data and process it on the fly.
That's said, approach when Processor hosts its own API instance looks cumbersome to me and doesn't provide any advantages.
